File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/CapsulePlacementStore.kt
package com.tabtin.mobile.features.conversation

import android.content.Context
import kotlin.math.max
import kotlin.math.min

internal enum class CapsuleDockSide {
    LEFT,
    RIGHT,
}

internal data class CapsulePlacement(
    val side: CapsuleDockSide = CapsuleDockSide.RIGHT,
    /** 0 = 可拖区域顶部，1 = 底部。 */
    val yRatio: Float = 1f,
) {
    internal companion object {
        val DEFAULT: CapsulePlacement = CapsulePlacement()
    }
}

internal data class CapsulePlacementBounds(
    val minX: Float,
    val maxX: Float,
    val minY: Float,
    val maxY: Float,
)

internal object CapsulePlacementMetrics {
    const val SAFE_MARGIN_PX: Float = 14f
    const val HARD_TOP_INSET_PX: Float = 0f
}

internal object CapsulePlacementGeometry {
    fun clamp(value: Float, minValue: Float, maxValue: Float): Float =
        max(minValue, min(maxValue, value))

    fun resolveBounds(
        viewportWidth: Float,
        viewportHeight: Float,
        capsuleWidth: Float,
        capsuleHeight: Float,
        safeMargin: Float = CapsulePlacementMetrics.SAFE_MARGIN_PX,
        hardTopInset: Float = CapsulePlacementMetrics.HARD_TOP_INSET_PX,
    ): CapsulePlacementBounds {
        val margin = max(0f, safeMargin)
        val hardMaxX = max(0f, viewportWidth - capsuleWidth)
        val hardMaxY = max(0f, viewportHeight - capsuleHeight)

        val preferredMinX = margin
        val preferredMaxX = viewportWidth - margin - capsuleWidth
        val (minX, maxX) = if (preferredMinX <= preferredMaxX) {
            clamp(preferredMinX, 0f, hardMaxX) to clamp(preferredMaxX, 0f, hardMaxX)
        } else {
            0f to hardMaxX
        }

        val preferredMinY = hardTopInset + margin
        val preferredMaxY = viewportHeight - margin - capsuleHeight
        val (minY, maxY) = if (preferredMinY <= preferredMaxY) {
            clamp(preferredMinY, 0f, hardMaxY) to clamp(preferredMaxY, 0f, hardMaxY)
        } else {
            val hardMin = if (hardMaxY >= hardTopInset) hardTopInset else 0f
            clamp(hardMin, 0f, hardMaxY) to hardMaxY
        }

        return CapsulePlacementBounds(minX, maxX, minY, maxY)
    }

    fun position(
        placement: CapsulePlacement,
        viewportWidth: Float,
        viewportHeight: Float,
        capsuleWidth: Float,
        capsuleHeight: Float,
        safeMargin: Float = CapsulePlacementMetrics.SAFE_MARGIN_PX,
        hardTopInset: Float = CapsulePlacementMetrics.HARD_TOP_INSET_PX,
    ): Pair<Float, Float> {
        val bounds = resolveBounds(
            viewportWidth,
            viewportHeight,
            capsuleWidth,
            capsuleHeight,
            safeMargin,
            hardTopInset,
        )
        val ratio = clamp(placement.yRatio, 0f, 1f)
        val x = if (placement.side == CapsuleDockSide.LEFT) bounds.minX else bounds.maxX
        val y = bounds.minY + (bounds.maxY - bounds.minY) * ratio
        return x to y
    }

    fun placement(
        x: Float,
        y: Float,
        viewportWidth: Float,
        viewportHeight: Float,
        capsuleWidth: Float,
        capsuleHeight: Float,
        safeMargin: Float = CapsulePlacementMetrics.SAFE_MARGIN_PX,
        hardTopInset: Float = CapsulePlacementMetrics.HARD_TOP_INSET_PX,
    ): CapsulePlacement {
        val bounds = resolveBounds(
            viewportWidth,
            viewportHeight,
            capsuleWidth,
            capsuleHeight,
            safeMargin,
            hardTopInset,
        )
        val clampedX = clamp(x, bounds.minX, bounds.maxX)
        val clampedY = clamp(y, bounds.minY, bounds.maxY)
        val midX = (bounds.minX + bounds.maxX) / 2f
        val rangeY = bounds.maxY - bounds.minY
        return CapsulePlacement(
            side = if (clampedX < midX) CapsuleDockSide.LEFT else CapsuleDockSide.RIGHT,
            yRatio = if (rangeY > 0f) (clampedY - bounds.minY) / rangeY else 1f,
        )
    }

    fun dockedPosition(
        x: Float,
        y: Float,
        viewportWidth: Float,
        viewportHeight: Float,
        capsuleWidth: Float,
        capsuleHeight: Float,
        safeMargin: Float = CapsulePlacementMetrics.SAFE_MARGIN_PX,
        hardTopInset: Float = CapsulePlacementMetrics.HARD_TOP_INSET_PX,
    ): Pair<Float, Float> {
        val snapped = placement(
            x,
            y,
            viewportWidth,
            viewportHeight,
            capsuleWidth,
            capsuleHeight,
            safeMargin,
            hardTopInset,
        )
        return position(
            snapped,
            viewportWidth,
            viewportHeight,
            capsuleWidth,
            capsuleHeight,
            safeMargin,
            hardTopInset,
        )
    }
}

internal class CapsulePlacementStore(context: Context) {
    private val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    fun load(): CapsulePlacement {
        val side = when (prefs.getString(KEY_SIDE, CapsuleDockSide.RIGHT.name)) {
            CapsuleDockSide.LEFT.name -> CapsuleDockSide.LEFT
            else -> CapsuleDockSide.RIGHT
        }
        val ratio = CapsulePlacementGeometry.clamp(prefs.getFloat(KEY_Y_RATIO, 1f), 0f, 1f)
        return CapsulePlacement(side, ratio)
    }

    fun save(placement: CapsulePlacement) {
        prefs.edit()
            .putString(KEY_SIDE, placement.side.name)
            .putFloat(KEY_Y_RATIO, placement.yRatio)
            .apply()
    }

    private companion object {
        private const val PREFS = "tt_workbench_capsule_placement"
        private const val KEY_SIDE = "side"
        private const val KEY_Y_RATIO = "yRatio"
    }
}

internal enum class CapsuleOnboardingAction {
    TAP,
    DRAG,
    HOLD,
}

internal data class CapsuleOnboardingProgress(
    val appearanceCount: Int = 0,
    val lastPromptAppearance: Int = 0,
    val tapLearned: Boolean = false,
    val dragLearned: Boolean = false,
    val holdLearned: Boolean = false,
    val skipped: Boolean = false,
) {
    fun recordAppearance(): CapsuleOnboardingProgress =
        copy(appearanceCount = appearanceCount + 1)

    fun nextPrompt(replySuggested: Boolean): CapsuleOnboardingAction? {
        if (skipped || lastPromptAppearance >= appearanceCount) return null
        if (!tapLearned) return CapsuleOnboardingAction.TAP
        if (replySuggested && !holdLearned && appearanceCount >= 2) {
            return CapsuleOnboardingAction.HOLD
        }
        if (!dragLearned && appearanceCount >= 2) return CapsuleOnboardingAction.DRAG
        return null
    }

    fun markPromptShown(): CapsuleOnboardingProgress =
        copy(lastPromptAppearance = appearanceCount)

    fun markLearned(action: CapsuleOnboardingAction): CapsuleOnboardingProgress = when (action) {
        CapsuleOnboardingAction.TAP -> copy(tapLearned = true)
        CapsuleOnboardingAction.DRAG -> copy(dragLearned = true)
        CapsuleOnboardingAction.HOLD -> copy(holdLearned = true)
    }

    fun skipAll(): CapsuleOnboardingProgress = copy(skipped = true)
}

internal class CapsuleOnboardingStore(context: Context) {
    private val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    fun load(): CapsuleOnboardingProgress = CapsuleOnboardingProgress(
        appearanceCount = prefs.getInt(KEY_APPEARANCES, 0),
        lastPromptAppearance = prefs.getInt(KEY_LAST_PROMPT_APPEARANCE, 0),
        tapLearned = prefs.getBoolean(KEY_TAP_LEARNED, false),
        dragLearned = prefs.getBoolean(KEY_DRAG_LEARNED, false),
        holdLearned = prefs.getBoolean(KEY_HOLD_LEARNED, false),
        skipped = prefs.getBoolean(KEY_SKIPPED, false),
    )

    fun save(progress: CapsuleOnboardingProgress) {
        prefs.edit()
            .putInt(KEY_APPEARANCES, progress.appearanceCount)
            .putInt(KEY_LAST_PROMPT_APPEARANCE, progress.lastPromptAppearance)
            .putBoolean(KEY_TAP_LEARNED, progress.tapLearned)
            .putBoolean(KEY_DRAG_LEARNED, progress.dragLearned)
            .putBoolean(KEY_HOLD_LEARNED, progress.holdLearned)
            .putBoolean(KEY_SKIPPED, progress.skipped)
            .apply()
    }

    fun reset() {
        prefs.edit().clear().apply()
    }

    private companion object {
        private const val PREFS = "tt_workbench_capsule_onboarding_v1"
        private const val KEY_APPEARANCES = "appearanceCount"
        private const val KEY_LAST_PROMPT_APPEARANCE = "lastPromptAppearance"
        private const val KEY_TAP_LEARNED = "tapLearned"
        private const val KEY_DRAG_LEARNED = "dragLearned"
        private const val KEY_HOLD_LEARNED = "holdLearned"
        private const val KEY_SKIPPED = "skipped"
    }
}

File: apps/tabtin-ios/Tabtin/Features/Workbench/CapsulePlacementStore.swift
import CoreGraphics
import Foundation

enum CapsuleDockSide: String, Equatable, Sendable {
    case left
    case right
}

struct CapsulePlacement: Equatable, Sendable {
    var side: CapsuleDockSide
    /// 0 = 可拖区域顶部，1 = 底部。
    var yRatio: CGFloat

    static let `default` = CapsulePlacement(side: .right, yRatio: 1)
}

enum CapsulePlacementMetrics {
    static let safeMargin: CGFloat = 14
    /// 额外预留给顶栏 / 导航的硬顶 inset（相对容器）。
    static let hardTopInset: CGFloat = 0
}

enum CapsulePlacementGeometry {
    struct Bounds: Equatable, Sendable {
        var minX: CGFloat
        var maxX: CGFloat
        var minY: CGFloat
        var maxY: CGFloat
    }

    static func clamp(_ value: CGFloat, _ min: CGFloat, _ max: CGFloat) -> CGFloat {
        Swift.max(min, Swift.min(max, value))
    }

    static func resolveBounds(
        viewport: CGSize,
        capsuleSize: CGSize,
        safeMargin: CGFloat = CapsulePlacementMetrics.safeMargin,
        hardTopInset: CGFloat = CapsulePlacementMetrics.hardTopInset
    ) -> Bounds {
        let margin = Swift.max(0, safeMargin)
        let hardMaxX = Swift.max(0, viewport.width - capsuleSize.width)
        let hardMaxY = Swift.max(0, viewport.height - capsuleSize.height)

        let preferredMinX = margin
        let preferredMaxX = viewport.width - margin - capsuleSize.width
        let minX: CGFloat
        let maxX: CGFloat
        if preferredMinX <= preferredMaxX {
            minX = clamp(preferredMinX, 0, hardMaxX)
            maxX = clamp(preferredMaxX, 0, hardMaxX)
        } else {
            minX = 0
            maxX = hardMaxX
        }

        let preferredMinY = hardTopInset + margin
        let preferredMaxY = viewport.height - margin - capsuleSize.height
        let minY: CGFloat
        let maxY: CGFloat
        if preferredMinY <= preferredMaxY {
            minY = clamp(preferredMinY, 0, hardMaxY)
            maxY = clamp(preferredMaxY, 0, hardMaxY)
        } else {
            let hardMin = hardMaxY >= hardTopInset ? hardTopInset : 0
            minY = clamp(hardMin, 0, hardMaxY)
            maxY = hardMaxY
        }

        return Bounds(minX: minX, maxX: maxX, minY: minY, maxY: maxY)
    }

    static func position(
        for placement: CapsulePlacement,
        viewport: CGSize,
        capsuleSize: CGSize,
        safeMargin: CGFloat = CapsulePlacementMetrics.safeMargin,
        hardTopInset: CGFloat = CapsulePlacementMetrics.hardTopInset
    ) -> CGPoint {
        let bounds = resolveBounds(
            viewport: viewport,
            capsuleSize: capsuleSize,
            safeMargin: safeMargin,
            hardTopInset: hardTopInset
        )
        let ratio = clamp(placement.yRatio, 0, 1)
        return CGPoint(
            x: placement.side == .left ? bounds.minX : bounds.maxX,
            y: bounds.minY + (bounds.maxY - bounds.minY) * ratio
        )
    }

    static func placement(
        from position: CGPoint,
        viewport: CGSize,
        capsuleSize: CGSize,
        safeMargin: CGFloat = CapsulePlacementMetrics.safeMargin,
        hardTopInset: CGFloat = CapsulePlacementMetrics.hardTopInset
    ) -> CapsulePlacement {
        let bounds = resolveBounds(
            viewport: viewport,
            capsuleSize: capsuleSize,
            safeMargin: safeMargin,
            hardTopInset: hardTopInset
        )
        let clamped = CGPoint(
            x: clamp(position.x, bounds.minX, bounds.maxX),
            y: clamp(position.y, bounds.minY, bounds.maxY)
        )
        let midX = (bounds.minX + bounds.maxX) / 2
        let rangeY = bounds.maxY - bounds.minY
        return CapsulePlacement(
            side: clamped.x < midX ? .left : .right,
            yRatio: rangeY > 0 ? (clamped.y - bounds.minY) / rangeY : 1
        )
    }

    /// 松手吸附：水平贴最近边，竖直保持当前 y。
    static func dockedPosition(
        from position: CGPoint,
        viewport: CGSize,
        capsuleSize: CGSize,
        safeMargin: CGFloat = CapsulePlacementMetrics.safeMargin,
        hardTopInset: CGFloat = CapsulePlacementMetrics.hardTopInset
    ) -> CGPoint {
        let snapped = placement(
            from: position,
            viewport: viewport,
            capsuleSize: capsuleSize,
            safeMargin: safeMargin,
            hardTopInset: hardTopInset
        )
        return self.position(
            for: snapped,
            viewport: viewport,
            capsuleSize: capsuleSize,
            safeMargin: safeMargin,
            hardTopInset: hardTopInset
        )
    }
}

enum CapsuleHITLBubbleGeometry {
    enum Edge: Equatable, Sendable {
        case above
        case below
    }

    struct Placement: Equatable, Sendable {
        let frame: CGRect
        let edge: Edge
    }

    static let gap: CGFloat = 8
    static let idealWidth: CGFloat = 288

    static func maximumSize(
        viewport: CGRect,
        capsuleFrame: CGRect,
        safeMargin: CGFloat = CapsulePlacementMetrics.safeMargin
    ) -> CGSize {
        let margin = max(0, safeMargin)
        let safeBounds = viewport.insetBy(dx: margin, dy: margin)
        let roomAbove = max(0, capsuleFrame.minY - gap - safeBounds.minY)
        let roomBelow = max(0, safeBounds.maxY - capsuleFrame.maxY - gap)
        return CGSize(
            width: min(idealWidth, max(0, safeBounds.width)),
            height: max(roomAbove, roomBelow)
        )
    }

    /// 独立计算气泡 frame；胶囊 frame 只作为锚点，不参与 `CapsulePlacementStore` 的
    /// chrome 测量，因此气泡出现/消失不会改写 yRatio 或让胶囊跳位。
    static func placement(
        viewport: CGRect,
        capsuleFrame: CGRect,
        bubbleSize: CGSize,
        side: CapsuleDockSide,
        safeMargin: CGFloat = CapsulePlacementMetrics.safeMargin
    ) -> Placement {
        let margin = max(0, safeMargin)
        let safeBounds = viewport.insetBy(dx: margin, dy: margin)
        let maximumSize = maximumSize(
            viewport: viewport,
            capsuleFrame: capsuleFrame,
            safeMargin: safeMargin
        )
        let constrainedSize = CGSize(
            width: min(max(0, bubbleSize.width), maximumSize.width),
            height: min(max(0, bubbleSize.height), maximumSize.height)
        )
        let alignedX = side == .left
            ? capsuleFrame.minX
            : capsuleFrame.maxX - constrainedSize.width
        let minX = safeBounds.minX
        let maxX = max(minX, safeBounds.maxX - constrainedSize.width)
        let x = CapsulePlacementGeometry.clamp(alignedX, minX, maxX)

        let aboveY = capsuleFrame.minY - gap - constrainedSize.height
        let belowY = capsuleFrame.maxY + gap
        let roomAbove = capsuleFrame.minY - gap - safeBounds.minY
        let roomBelow = safeBounds.maxY - capsuleFrame.maxY - gap
        let edge: Edge = roomAbove >= roomBelow
            ? .above
            : .below
        let desiredY = edge == .above ? aboveY : belowY
        let minY = safeBounds.minY
        let maxY = max(minY, safeBounds.maxY - constrainedSize.height)
        let y = CapsulePlacementGeometry.clamp(desiredY, minY, maxY)

        return Placement(
            frame: CGRect(origin: CGPoint(x: x, y: y), size: constrainedSize),
            edge: edge
        )
    }
}

/// 气泡刚插入 SwiftUI 树时，真实尺寸偏好可能要到下一帧才上报。
/// 先用一个安全的卡片尺寸参与定位，避免“尺寸为零 → 透明 → 永远测不到”的自锁。
enum CapsuleHITLAccessoryMeasurement {
    static let bootstrapHeight: CGFloat = 180

    static func resolvedSize(measured: CGSize, maximum: CGSize) -> CGSize {
        guard maximum.width > 1, maximum.height > 1 else { return .zero }
        if measured.width > 1, measured.height > 1 {
            return CGSize(
                width: min(measured.width, maximum.width),
                height: min(measured.height, maximum.height)
            )
        }
        return CGSize(
            width: maximum.width,
            height: min(bootstrapHeight, maximum.height)
        )
    }
}

/// 按设备本地持久化胶囊贴边位置（side + yRatio）。
@MainActor
enum CapsulePlacementStore {
    private static let defaultsKey = "tt.workbench.capsulePlacement"

    static func load() -> CapsulePlacement {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return .default }
        let sideRaw = obj["side"] as? String
        let side = CapsuleDockSide(rawValue: sideRaw ?? "") ?? .right
        let ratioValue = obj["yRatio"] as? Double ?? Double(CapsulePlacement.default.yRatio)
        let ratio = CapsulePlacementGeometry.clamp(CGFloat(ratioValue), 0, 1)
        return CapsulePlacement(side: side, yRatio: ratio)
    }

    static func save(_ placement: CapsulePlacement) {
        let payload: [String: Any] = [
            "side": placement.side.rawValue,
            "yRatio": Double(placement.yRatio),
        ]
        if let data = try? JSONSerialization.data(withJSONObject: payload) {
            UserDefaults.standard.set(data, forKey: defaultsKey)
        }
    }
}

enum CapsuleOnboardingAction: String, Codable, Equatable, Sendable {
    case tap
    case drag
    case hold
}

/// 胶囊教学只记录设备本地的发现进度，不跟账号或工作区绑定。
struct CapsuleOnboardingProgress: Codable, Equatable, Sendable {
    var appearanceCount = 0
    var lastPromptAppearance = 0
    var tapLearned = false
    var dragLearned = false
    var holdLearned = false
    var skipped = false

    mutating func recordAppearance() {
        appearanceCount += 1
    }

    func nextPrompt(replySuggested: Bool) -> CapsuleOnboardingAction? {
        guard !skipped, lastPromptAppearance < appearanceCount else { return nil }
        if !tapLearned { return .tap }
        if replySuggested, !holdLearned, appearanceCount >= 2 { return .hold }
        if !dragLearned, appearanceCount >= 2 { return .drag }
        return nil
    }

    mutating func markPromptShown(_: CapsuleOnboardingAction) {
        lastPromptAppearance = appearanceCount
    }

    mutating func markLearned(_ action: CapsuleOnboardingAction) {
        switch action {
        case .tap: tapLearned = true
        case .drag: dragLearned = true
        case .hold: holdLearned = true
        }
    }

    mutating func skipAll() {
        skipped = true
    }
}

@MainActor
enum CapsuleOnboardingStore {
    private static let defaultsKey = "tt.workbench.capsuleOnboarding.v1"

    static func load() -> CapsuleOnboardingProgress {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey),
              let progress = try? JSONDecoder().decode(CapsuleOnboardingProgress.self, from: data)
        else { return CapsuleOnboardingProgress() }
        return progress
    }

    static func save(_ progress: CapsuleOnboardingProgress) {
        guard let data = try? JSONEncoder().encode(progress) else { return }
        UserDefaults.standard.set(data, forKey: defaultsKey)
    }

    static func reset() {
        UserDefaults.standard.removeObject(forKey: defaultsKey)
    }
}
